### PR TITLE
Automatic PR creation for yosys-symbiflow-plugins update

### PR DESCRIPTION
**Pull request by coolbreeze413**
> ### Motivate of the pull request
> - [ ] To address an existing issue. If so, please provide a link to the issue: <issue id>
> - [ ] Breaking new feature. If so, please describe details in the description part.

Use dependabot to create PRs automatically when yosys-symbiflow-plugins is updated.

> ### Describe the technical details
> #### What is currently done? (Provide issue link if applicable)
> 
> 
> 
> 
> 
>
> #### What does this pull request change?
> 
> 
> 
> 
> 

Adds dependabot yaml configuration to update yosys-plugins

> ### Which part of the code base require a change
> 
> - [ ] VPR
> - [ ] Tileable routing architecture generator
> - [ ] OpenFPGA libraries
> - [ ] FPGA-Verilog
> - [ ] FPGA-Bitstream
> - [ ] FPGA-SDC
> - [ ] FPGA-SPICE
> - [ ] Flow scripts
> - [ ] Architecture library
> - [ ] Cell library
> - [ ] Documentation
> - [ ] Regression tests
> - [ ] Continous Integration (CI) scripts

> ### Impact of the pull request

> - [ ] Require a change on Quality of Results (QoR)
> - [ ] Break back-compatibility. If so, please list who may be influenced.

### CODE REVIEW

**Comment by coolbreeze413** [comment]
At this point, this is a draft PR, will mark as ready once I do a separate test of dependabot for specific submodules.
Please review the initial configuration and let me know if you see any obvious issues.

**Comment by coolbreeze413** [comment]
I have tested this with a couple of experimental repos and it looks good to go.
The current configuration will ensure that only the specific submodule mentioned in `allow` will be checked and PR created, in this case only `yosys-plugins` (from the `.gitmodules` file)

By default, Dependabot checks for new versions at a random set time for the repository.
Hence set the schedule to 00:00 Mountain Time.

@tangxifan @ganeshgore Please review and let me know if you would like any other changes.

File: .github/dependabot.yml
Comment (tangxifan) [inline .github/dependabot.yml:15]: @coolbreeze413 Does it mean that it will create a PR without a default assignee and reviewer?

In general, it looks o.k. to me. @ganeshgore If you are fine, I propose to merge and see how it works after a few days.
@coolbreeze413 Are you fine to create any follow-up PR if we detect any issues?

**Review by tangxifan — APPROVED** [review]

File: .github/dependabot.yml
Comment (coolbreeze413) [reply .github/dependabot.yml:15]: @tangxifan  yes the PR will not have any assignee or reviewer.
This can be set if you think that is better.

You can check an example PR here from my experiments: https://github.com/coolbreeze413/dependabot_main/pull/3

It can be merged by the bot by leaving a `@dependabot merge` comment as I tried here, or it can be manually merged.
Once the latest PR is merged, any previous unmerged PRs for the same submodule will be closed automatically as well, for example: https://github.com/coolbreeze413/dependabot_main/pull/2

Also, the bot will add a default `dependencies` label, if needed a custom one can be set as well.

I think it is a good idea to try it out, and then adjust the configuration as needed, I can raise PRs for the same accordingly, no problems.

Comment (coolbreeze413) [reply .github/dependabot.yml:15]: @tangxifan here is an example PR with the assignee and reviewer assigned automatically: https://github.com/coolbreeze413/dependabot_main/pull/4

**Comment by tangxifan** [comment]
@coolbreeze413 Just a few question
- Will the auto merge consider CI green as a requirement? I personally does not like auto-merge because it may bury some problems. A double check is safer.
- I am not sure what criteria the dependabot will be triggered. For example, each time there is an update on the master/main branch of a submodule or only when there is a release? If it is a release, a daily schedule is o.k. otherwise we may consider weekly. My concern is that it will create too many PRs automatically (spamming our PR list)
- If we choose not to auto merge, can you use dependabot to assign me and @tpagarani as reviewers?

**Comment by coolbreeze413** [comment]
@tangxifan please find my anwers below:

- dependabot will not auto-merge. 

  The merge will be done only if we comment `@dependabot merge` and then, only if the other criteria is still met i.e in our case,     at least 1 approving review, and all the CI is green.

  WIthout that, the PR cannot be merged: https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/managing-pull-requests-for-dependency-updates

- dependabot will raise PR anytime the version of the submodule's repo is ahead of the version here, however the maximum PRs that will be open at anytime can be set (it is 5 by default). 
  
  Older ones are closed automatically.

- I will add you and @tpagarani  as reviewers. Do you want a default assignee as well?

**Comment by ganeshgore** [comment]
@coolbreeze413
This looks good, here are my thoughts 

1. I am ok if you

**Comment by ganeshgore** [comment]
@coolbreeze413 
This looks good. here are my comments. 

1. I am ok if this setup creates PR and auto-merges after CI is green, and PR remains open if CI fails (I don't think If CI is green then there is anything to review from our side)
2. About the multiple numbers of PR, my understanding is, at any given point there will be either no pending PR (if CI is green) or only one PR (if CI failed)
3. If `yosys-plugins` development happens directly on the master branch, then we can stick to once in a day + manually triggered dependable runs

let me what you guys think

**Comment by tangxifan** [comment]
@ganeshgore I am totally o.k. that if PR get auto-merged as long as CI is green. We have limited bandwidth on code review. For frequency, once a day may be too aggressive and a weekly update looks reasonable. Anyway, we do not know what will happen in practice. If we converge and merge this PR, then we keep monitoring a few days and see if any critical issue raises.

**Comment by coolbreeze413** [comment]
@tangxifan @ganeshgore : In answer to your comments:

- For enabling auto-merge (and auto-review):

   We would need to create Github Actions workflow to do this using the PR created, this explains the steps: https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/automating-dependabot-with-github-actions#enable-auto-merge-on-a-pull-request

  Also, the merge will happen only if the criteria already specified for any PR is met as usual : CI should be green, an approving review (auto-review can be done via Github Actions as well)

  This will take a couple of days for me to test/experiment on, and make an appropriate Github Actions workflow.

  If you are ok to keep auto-merge as a separate PR, we can enable the bot upto creating the PR automatically for now, merging remains manual.

- Number of Pending PRs:

  The bot will keep opening new PRs whenever the submodule version is updated.
  The bot will stop creating new PRs when the number of already open PRs == open-pull-requests-limit
  Once some PRs are closed/merged, then the bot will start creating PRs again.
   https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/configuration-options-for-dependency-updates#open-pull-requests-limit

- Interval:

  I think it is good to go with a `daily` interval to begin with and then adjust it according to the frequency of updates in the yosys-symbiflow-repo.

  Also, if auto-merge is enabled, then I think `daily` interval should be ok.

**Comment by tangxifan** [comment]
@coolbreeze413 Make sense to me. We can merge while you are working on auto-merge features.

**Review by ganeshgore — APPROVED** [review]

**Comment by tangxifan** [comment]
Waiting for @tpagarani and @lnsharma 's final confirmation.

**Comment by tpagarani** [comment]
LGTM

**Review by tpagarani — APPROVED** [review]

**Pull request merged**
